File: Strings/first uniqu character/C/unique_char.c
#include <stdio.h>
#include <string.h>




int firstUniqChar(char * s){
    
    /* evevry element of hash array is letter[0|1|2_is_it_known|unique|duplicated, first_index_in_string_s] */
    unsigned int hash[26][2] = {0};
    
    // create hash table
    for (unsigned int index = 0; index < strlen(s); index++){
        unsigned char pos = (int) s[index] - 97; // ascii 'a' = 97
        if (hash[pos][0] == 2){ 
            continue;
        }
        else if (hash[pos][0] == 0){
            hash[pos][0] = 1; // unique from unknown
            hash[pos][1] = index; // position of potentionaly unique char
        }
       
        else if (hash[pos][0] == 1){
            hash[pos][0] = 2;
        }
    }
    
    // go trough hash table
    unsigned int result = strlen(s);
    for (unsigned char i = 0; i < 26; i++){
        if (hash[i][0] == 1){
            if (hash[i][1] < result){
                result = hash[i][1];
            }
        }
    }

    if (result == strlen(s)){
        return -1;
    }
    else{
        return result;
    }
}

int main(void){
    char *test_s = "aabb";
    
    
    printf("\nfirst unique char position is %i\n", firstUniqChar(test_s));
    
}

File: Strings/first uniqu character/JS/unique_char.js
/**
 * @param {string} s
 * @return {number}
 */
 var firstUniqChar = function(s) {
    const hash_table = new Map();

    /* create hash table key - alpha letter, \
    value - list [is_unique, first_index]*/
    for (let i = 0; i < s.length; i++){
        let letter = s.charAt(i);
        if (hash_table.has(letter)){
            hash_table.get(letter)[0] = false;
        }
        else{
            hash_table.set(letter, [true, i]);
        }
    }
    /* walk trough hash table to look for true unique elements in ordered dict map*/
    for (let value of hash_table.values()){
        if (value[0] === true){
            return value[1];
        }
    }
    return -1;
};

var firstUniqChar_v2 = function(s) {
    
    let uniqueChars = new Map();
    
    for(let i = 0; i < s.length; i++) {
        if(uniqueChars.has(s[i])) {
            uniqueChars.get(s[i]).push(i);
        } else {
            uniqueChars.set(s[i], [i]);
        }
    }
    
    
    for(const [key, value] of uniqueChars) {
        
        if(value.length == 1) {
          return value[0];
        } 
    }
    
    return -1;
    
};

let test_s = "aabb";
console.log(`first unique char in "${test_s}" is ${firstUniqChar(test_s)}`);